Remove compiler and use nimscript with nim e instead

genotrance opened the pull request:
- Removed compiler dependency and use `nim e` instead
- Added `nimscriptwrapper.nim` to replace `nimscriptsupport.nim` (still in source control)
- Added OSX to CI
- Added devel & stable to CI
- Tests fixed to pass on Windows

genotrance commented:
cc @Araq @dom96

Araq commented:
Ready to review?

genotrance commented:
Yes good to go, passes all tests on Linux, OSX and Windows, both on stable and devel.

Araq commented on nimble.nimble:
```diff
@@ -1,7 +1,9 @@
 
-when fileExists(thisModuleFile.parentDir / "src/nimblepkg/common.nim"):
+const
+  parentDir = currentSourcePath.rsplit(seps={'/', '\\', ':'}, maxsplit=1)[0]
```
Why is ':' listed here?

genotrance replied:
Poor man's path split. Will all go away since @dom96 wants to hard code the version. This is causing another issue when nimble is installed as a package.

genotrance replied:
os.nim cannot be imported in nimscript.

Araq replied:
It can with Nim devel and before that there was ospaths.nim

genotrance replied:
Removed all this code and replaced version = "0.9.0".

Araq commented on src/nimblepkg/nimscriptapi.nim:
```diff
@@ -22,46 +25,157 @@ var
 
+proc getParams() =
+  for i in 5 .. paramCount():
+    if paramStr(i).len != 0:
```
``paramStr(i).len != 0`` should always be true.

genotrance replied:
This was what caused the issue I mentioned. Empty string was coming in as a param. In fixed it in the caller so can be removed.

genotrance replied:
Done

Araq commented:
LGTM.

dom96 commented on src/nimblepkg/nimscriptwrapper.nim:
```diff
@@ -0,0 +1,131 @@
+  result.iniFile = prjCacheDir / scriptName.extractFilename().changeFileExt ".ini"
+
+  if not prjCacheDir.dirExists() or not nimsCacheFile.fileExists() or not result.iniFile.fileExists() or
+    scriptName.getLastModificationTime() > nimsCacheFile.getLastModificationTime():
```
```suggestion
    if not isScriptResultCached:
```

dom96 commented on src/nimblepkg/nimscriptwrapper.nim:
```diff
@@ -0,0 +1,131 @@
+  result.nimsFile = scriptName.parentDir() / scriptName.splitFile().name & "_" & shash & ".nims"
+  result.iniFile = prjCacheDir / scriptName.extractFilename().changeFileExt ".ini"
+
+  if not prjCacheDir.dirExists() or not nimsCacheFile.fileExists() or not result.iniFile.fileExists() or
```
```suggestion
  let isScriptResultCached =
    prjCacheDir.dirExists() and nimsCacheFile.fileExists() and
    result.iniFile.fileExists() and 
    scriptName.getLastModificationTime() > nimsCacheFile.getLastModificationTime()
```

dom96 commented on src/nimblepkg/nimscriptwrapper.nim:
```diff
@@ -0,0 +1,131 @@
+    writeFile(nimsCacheFile, nimscriptApi & scriptName.readFile() & "\nonExit()\n")
+    discard tryRemoveFile(result.iniFile)
+
+  if not result.nimsFile.fileExists():
```
huh? Why would this not exist? I'm assuming that `result.nimsFile` is the nims file that is being executed, i.e. the .nimble file in my package.

genotrance replied:
The nims file gets generated once and saved in nimblecache and updated if the nimble file changes. When tasks need to be executed, it gets copied into the folder next to the nimble file and executed there. This is because all existing nimble scripts assume they will be in that location, either to import files or access relative paths.

However, we cannot leave the file sitting there since it is unexpected and will stand out with git status, etc. So it is deleted at the end of each call.

dom96 replied:
Right. I would prefer if a single function dealth with this temporary `nims` file, instead of returning it and having another function delete it. I think with the suggestion I gave above this would be possible and make the code more contained.

genotrance replied:
The nims file copy and delete is done in the same proc now.

dom96 commented on src/nimblepkg/nimscriptwrapper.nim:
```diff
@@ -0,0 +1,131 @@
+  else:
+    result = execCmdEx(cmd, options = {poUsePath})
+
+proc setupNimscript*(scriptName: string, options: Options): tuple[nimsFile, iniFile: string] =
```
Nit: This function's name is misleading, it doesn't _just_ setup the Nimscript, it also executes under certain circumstances. This should be reflected in the name.

I know you're caching the result, but it still feels wrong to always execute `pringPkgInfo` before running a task. I would put this logic in a separate procedure.

genotrance replied:
Generating the nims file and saving the ini file is part of the setup. setupNimscript is kind of like setupVM in that sense. The execution is required, without running `nim e printPkgInfo`, nimble won't have package info. That's why it is bundled into the setup call.

dom96 replied:
setupVM doesn't retrieve the package info though. Package info is only required when Nimble asks for it, when you're executing a task it isn't needed so why read it?

I would prefer to see this organised like this:

* executeNimscript(scriptName, action: string, options: Options): ExecutionResult[string]
* retrievePackageInfo(scriptName: string, options: Options): ExecutionResult[string]
  * This is where you call ``executeNimscript(scriptName, "printPkgInfo", options)`` and have all the .ini reading logic.

Does that make sense?

genotrance replied:
I've cleaned this up - setupNimscript() only creates the nims and nimscriptapi.nim files and is not invoked outside the file. getIniFile() is the only proc exposed (apart from execHook and execTask) and creates the ini file and returns the path.

dom96 commented on src/nimblepkg/nimscriptwrapper.nim:
```diff
@@ -0,0 +1,131 @@
+
+  result.flags = newTable[string, seq[string]]()
+  if "success" in j:
+    result.success = j["success"].getBool()
```
These can be simplified to:

```nim
result.success = j{"success"}.getBool()
```

and the same for others.

dom96 replied:
You could actually probably just use the `to` macro here.

genotrance replied:
I'll defer this for now, don't want to spend more time on this, especially since there's a table in this object (flags).

genotrance commented on .travis.yml:
```diff
@@ -1,24 +1,24 @@
 
+env:
+  - BRANCH=0.19.4
+  - BRANCH=devel
```
We aren't tied to the compiler anymore, why does it need to be tied to a commit? If nimble fails to build due to compiler change, better to fail early right?

dom96 replied:
Because `devel` can break at any time, I don't want to get travis failures because of that.

genotrance replied:
Might it make sense to eventually pin to nightlies since those only get posted when all tests pass? Need to add the feature to choosenim to download/install nightlies but might make things more stable.

genotrance replied:
Done.

genotrance commented on .travis.yml:
```diff
@@ -1,24 +1,24 @@
   directories:
-    - "$HOME/.nimble"
-    - "$HOME/.choosenim"
+    - "$HOME/.choosenim/toolchains/nim-0.19.4"
```
There's nothing worth caching in ~/.nimble. Also, caching only stable since the binary won't change. Devel shouldn't be cached so that we always test with latest. Otherwise we are testing with an old devel after first cache.

dom96 replied:
> There's nothing worth caching in ~/.nimble. 

There is.

The `.nimble/bin` is added to the PATH below, if this directory doesn't get cached then failures will happen.

dom96 replied:
@genotrance as far as I'm concerned you need to revert this because it will break things, please do so.

genotrance commented on nimble.nimble:
```diff
@@ -1,7 +1,9 @@
+const
+  parentDir = currentSourcePath.rsplit(seps={'/', '\\', ':'}, maxsplit=1)[0]
+
+when fileExists(parentDir & "/src/nimblepkg/common.nim"):
```
I'll implement this in this pr itself if possible.

genotrance replied:
Using a define isn't feasible - nimble is compiled with `nim c` in koch so passing a define that's isolated to the nimble repo isn't possible.

For now, I'm going to hardcode 0.9.0 in both nimble.nimble and common.nim. If there's a better way to do this, I'll implement it separately for nimble and choosenim. Open to ideas.

genotrance commented on src/nimblepkg/nimscriptapi.nim:
```diff
@@ -22,46 +25,157 @@ var
+  else:
+    var
+      output = ""
+    output &= "\"success\": " & $success & ", "
```
Json doesn't work in nimscript. Even if it does in devel, it isn't backwards compatible and now that nimble is free of the compiler, there's value in making it backwards compatible.

dom96 replied:
That sucks. We should fix this and then change this code once it's fixed in a stable release.

genotrance commented on src/nimblepkg/nimscriptapi.nim:
```diff
@@ -22,46 +25,157 @@ var
+proc getParams() =
+  for i in 5 .. paramCount():
+    if paramStr(i).len != 0:
+      commandLineParams.add paramStr(i).normalize
```
Cannot even import parseopt since it conflicts with nimscript.

```
Error: ambiguous call; both system.paramCount() [declared in ..\..\..\..\Desktop\DL\programming\nimdevel\lib\system\nimscript.nim(65, 6)] and os.paramCount() declared in ..\..\..\..\Desktop\DL\programming\nimdevel\lib\pure\os.nim(2533, 8)] match for: ()
```

dom96 replied:
Again, ``import parseopt except paramCount``? or you could just disambiguate it here since this is your code

genotrance replied:
Nope - paramCount is exported by both system and nimscript, and imported by parseopt itself so the error is happening while loading that module. If the nims file is just as below, it will error. Nothing can be done except fixing parseopt to handle this scenario.

```nim
import parseopt
```

genotrance commented on src/nimblepkg/nimscriptapi.nim:
```diff
@@ -22,46 +25,157 @@ var
+  else:
+    flags[key] = @[value]
+
+template `--`(key, val: untyped) =
```
Long story but I started with `import nimscriptapi` but many of the global vars conflict with the instances in `system.nim` so nim complains to disambiguate. We cannot expect all nimble package authors to write `nimscriptapi.version`.

The only options are to `include nimscriptapi` or just have everything inline. Considering I was already checking and caching the nims files, I made nimscriptapi inline instead of having to check and write it separately. It also makes it easier to have multiple nimble processes running in parallel.

If there's a better way to get this working, I'm open to improve this.

dom96 replied:
Does `import system except version` not work? AFAIK we don't include this file right now, so this has to work somehow

genotrance replied:
I've implemented this but have to exclude every single nimscript var and task. Makes no sense.

```
import system except getCommand, setCommand, switch, `--`,
  packageName, version, author, description, license, srcDir, binDir, backend,
  skipDirs, skipFiles, skipExt, installDirs, installFiles, installExt, bin, foreignDeps,
  requires, task, packageName
```

genotrance commented on src/nimblepkg/nimscriptapi.nim:
```diff
@@ -22,46 +25,157 @@ var
     result = true
     body
 
+  beforeHooks.add astToStr(action)
```
This is used to communicate pre/post hooks back to nimble via the `printPkgInfo` method. It was being done in `nimscriptsupport.nim` but it is now being communicated back via the ini file. Nimble needs to know which pre/post hooks exist so that they can be called when appropriate.

dom96 replied:
I see. I forgot this was checked.

genotrance commented on src/nimblepkg/nimscriptwrapper.nim:
```diff
@@ -0,0 +1,131 @@
+  nimscriptApi = staticRead("nimscriptapi.nim")
+  nimscriptHash = $nimscriptApi.hash().abs()
+
+proc execNimscript(nimsFile, actionName: string, options: Options, live = true): tuple[output: string, exitCode: int] =
```
Cleaned up.

genotrance commented on src/nimblepkg/nimscriptwrapper.nim:
```diff
@@ -0,0 +1,131 @@
+  var
+    cmd = ("nim e --hints:off --verbosity:0 " & nimsFile.quoteShell & " " & actionName).strip()
+
+  if live:
```
So the live output lets the nimscript child process write to stdout directly so you get colored output and what not. But given there's no way to get the json content from stdout simultaneously (cannot read with poParentStreams and nimscript is limited), I had to resort to the .out file.

However, the printPkgInfo task writes to stdout and nimble then writes it to the ini file so no live output for that call. I can use actionName == "printPkgInfo" instead but wanted to keep execNimscript generic. Either is fine.

Am still wondering about using json for the ini file but will comment on that separately.

dom96 replied:
> I can use actionName == "printPkgInfo" instead but wanted to keep execNimscript generic. 

A way to make this even more generic would be to not special case this and to instead put the .ini file inside of ``retVal`` in the JSON. This would also allow you to return a typed ``ExecutionResult[string]`` from this procedure instead of the current ``string`` which can contain two different formats based on the value of `live`.

> But given there's no way to get the json content from stdout simultaneously (cannot read with poParentStreams and nimscript is limited)

I thought the plan was to output JSON amongst ordinary output and eventually deprecate the ability for Nimscript to echo, opting instead to use some sort of Nimble API. We can work on this aspect alter though, I don't mind :)

genotrance replied:
I don't see how the returne value matters - in one case, the output is collected from a file since stdout isn't accessible, in another case, it is piped.

As for JSON to stdout, didn't realize you didn't want to show what was happening. I initially was piping everything but you could have long running processes that are building in the background and the user won't get any feedback. We can figure that out separately though.

genotrance commented on src/nimblepkg/packageparser.nim:
```diff
@@ -259,6 +259,12 @@ proc readPackageInfoFromNimble(path: string; result: var PackageInfo) =
+              result.preHooks.incl(i.normalize)
+          of "afterhooks":
+            for i in ev.value.multiSplit:
+              result.postHooks.incl(i.normalize)
```
Any reason why you don't want to reuse existing code? I'll have to generate and json instead of the ini so it is pretty much the same in effect.

genotrance commented:
I have made most of the changes as suggested. If there aren't any major concerns, I'd prefer merging this as is so that it can undergo some general testing by the community before 0.20 comes out.

dom96 commented on nimble.nimble:
```diff
@@ -1,16 +1,6 @@
 # Package
 
-version       = nimbleVersion
+version       = "0.9.0"
```
👑

genotrance commented:
> @genotrance as far as I'm concerned you need to revert this because it will break things, please do so (or explain why it must be this way).

See [this job log](https://travis-ci.org/genotrance/nimble/jobs/524797558#L433) for starters. Cache is restored but install is still rerun. Travis docs don't say that restoring cache will skip the install step, neither in the [install](https://docs.travis-ci.com/user/job-lifecycle/) nor [cache](https://docs.travis-ci.com/user/caching/) documentation.

Even in nightlies we have to check for [existence](https://github.com/nim-lang/nightlies/blob/master/.travis.yml#L95) of nim binary to leverage caching.

Meanwhile, I tried caching the entire [~/.choosenim dir](https://travis-ci.org/genotrance/nimble/builds/524793107/config) but ran into random Travis issues so backed that out for now. It can be done later as we get Windows CI working, once [this PR](https://github.com/nimble-test/issue280and524/pull/2) is merged.

As for `~/.nimble`, no doubt choosenim lives there but the redownload and nimble setup takes 1.2 seconds. It is likely it is already cached in the Travis http proxy. This way, we also get newer versions when they get released. Also, most tests are run against a custom nimbleDir but not [generate](https://github.com/nimble-test/issue280and524/blob/master/issue280and524.nimble#L14) so if for whatever reason that test fails and an older generate binary is in ~/.nimble/bin, the test might still pass.

Overall, caching isn't material to this PR. I will add Windows CI and ~/.choosenim caching separately. Please push this through and I can proceed onto other tasks.

dom96 commented:
Alright. I forgot that `install` was run anyway in Travis.

So I've checked out your branch and tested it. It looks like the caching mechanism doesn't work, it never gets invalidated. I tried to modify the Nimble file version from 0.9.0 to 0.10.0 and  `nimble dump` keeps returning `0.9.0`. 

Then I removed the cache in `/tmp/nimblecache` and got this:

<img width="1193" alt="Screen Shot 2019-04-28 at 12 53 02" src="https://user-images.githubusercontent.com/246651/56863949-d16b2300-69b4-11e9-9c79-556b27093633.png">

Seems there is still some work left to do :/

genotrance commented:
> So I've checked out your branch and tested it. It looks like the caching mechanism doesn't work, it never gets invalidated. I tried to modify the Nimble file version from 0.9.0 to 0.10.0 and `nimble dump` keeps returning `0.9.0`.

Looks like I broke cache checking when I re-architected this. I've put in a fix and added a test case as well.

> Then I removed the cache in `/tmp/nimblecache` and got this:

I wasn't able to reproduce this.

dom96 commented:
Merged manually.

andreaferretti commented:
Apparently this PR broke **all** my packages tests :-D

All my libraries define the tasks `nimble tests` (a convention suggested in the [nimscript docs](https://nim-lang.org/docs/nims.html) ), and `nimble test` as an alias for that, since I usually forget the final s. For instance, in [patty nimble file](https://github.com/andreaferretti/patty/blob/master/patty.nimble#L12-L23) one can find

```nim
task tests, "run tests":
  --hints: off
  --linedir: on
  --stacktrace: on
  --linetrace: on
  --debuginfo
  --path: "."
  --run
  setCommand "c", "test.nim"

task test, "run tests":
  setCommand "tests"
```

After this PR, if I run `nimble test` on patty I get

```
andrea$ nimble test
  Executing task test in /Users/andrea/esperimenti/patty/patty.nimble
tests        run tests
test        run tests
   Warning: No tests found!
```

while running `nimble tests` gives an error

```
andrea$ nimble tests
  Executing task tests in /Users/andrea/esperimenti/patty/patty.nimble
../../../lib/pure/parsejson.nim(522) raiseParseErr
Error: unhandled exception: input(1, 153) Error: ] expected [JsonParsingError]
```

The same happens for all my other libraries. :-(

genotrance commented:
@andreaferretti - I've made a fix in #642.

andreaferretti commented:
@genotrance Thank you, now everything works again :-)

andreaferretti commented:
@genotrance Actually, I still find a strange issue in [Rosencrantz](https://github.com/andreaferretti/rosencrantz). Since the tests require a server and a client, I have a small [shell script](https://github.com/andreaferretti/rosencrantz/blob/master/test.sh) to launch both and stop the server after the tests are done:

```bash
set -e
set -u

nimble server
tests/rosencrantz &
PID="$!"
nimble client
kill "$PID"
```

It seems that the tests pass when launched manually. When I use the script, the tests do not pass (the returned value is not 0) and I get this strange line at the end:

```
(1, 22) Error: undeclared identifier: 'Users'
     Error: {"success": true, "command": "e", "flags": {"forceBuild": [""]}, "retVal": true}
```

which is even weirder cosindering that neither the server nor the client code contains the `Users` identifier anywhere

andreaferretti commented:
@genotrance I get random issue while testing *again* :-(

For instance, on [patty](https://github.com/andreaferretti/patty) I have the following definition

```nim
# patty.nimble

task tests, "run tests":
  --hints: off
  --linedir: on
  --stacktrace: on
  --linetrace: on
  --debuginfo
  --path: "."
  --run
  setCommand "c", "test.nim"

task test, "run tests":
setCommand "tests"
```

If I run `nimble tests` everything works, but if I run `nimble test` (no s), which should be an alias, I get

```
djerzinski:patty andrea$ nimble test
  Executing task test in /Users/andrea/esperimenti/patty/patty.nimble
tests        run tests
test        run tests
   Warning: No tests found!
```

If I invert the definition of `nimble test` and `nimble tests`, **both** commands fail in the same way. Same thing happens with [alea](https://github.com/unicredit/alea).

This is starting to get brittle. I have automated tests for my libraries, but I just cannot trust them anymore :-(

dom96 commented:
I noticed something similar for one of my tasks called ``server``, if I renamed it to `fooserver` it worked, but for some reason `server` didn't (other tasks in the same file called things like `client` also worked).

We need to fix this and release a hotfix Nimble before Nim 0.20.0 is out

genotrance commented:
I'll look at these issues asap.

genotrance commented:
I've figured out the bug - it is fragile since I'm avoiding `projectPath()` recently added in Nim devel. I did that so this new nimble could support v0.19.x as well. I'll make it more robust tomorrow.

dom96 commented:
What are you using instead?

genotrance commented:
I've created a fix - can you please verify on your end that it fixes your issues?

https://github.com/nim-lang/nimble/pull/659

andreaferretti commented:
Sure, but I'm away from computers now, I can check on Monday

andreaferretti commented:
@genotrance Works for me, thank you!
